### ISSUE TITLE
RouterRequest Write Method - Application Level Encryption

### ISSUE THREAD

**Issue by KaanYT** [IBM-Swift/Kitura#1461]
## Context and Description
We want to use application level encryption. We couldn't find proper way to do it without disabling Kitura's some features.

### Similar Example - JWT

It might look similar to [JWT](https://github.com/IBM-Swift/Swift-JWT/blob/master/Sources/SwiftJWT/JWTDecoder.swift). But it is not suitable for this case. 
* Diffrences
	* Keys are created for each session.
	* Keys are kept in session.
* Accessing session in Body Decoder is not possible.

### Discussed Solutions 
1) TypeSafeMiddleware
	* TypeSafeMiddleware for Codable routes that created the structure you are interested in and use a dummy struct for the body.
	* Disadvantages :
        * It looks like I can not use same TypeSafeMiddleware for different models.
            * We have to create TypeSafeMiddleware for each Model or I have to decode models in Routes.
2) Raw Routes
	* Create a normal middleware for raw routes that attaches the decrypted body to the request UserInfo.
	* Disadvantages :
        * Unnecessary weight in Request.
        * I have to downcase(From any) objects or use json decoder in each Routes.

### Expected vs. Actual Behaviour
- [ ] In Middleware, I should able to write RouterRequest before it is processed by Route.
- [ ] There is no method for write.
- [x] In Middleware, I should able to write RouterResponse after it is processed by Route. - (WrittenDataFilter) [Example - Kitura Compression](https://github.com/IBM-Swift/Kitura-Compression/blob/master/Sources/KituraCompression/Compression.swift)

#### Example Request Body
```json
Encrypted Request :
{
    "status":10,
    "encryptedText":"asdwqe823weafcxzblo"
}
Decrypted Request :
{
    "UserId":123,
    "UserName":"Test"
}
```

**Comment by djones6** [comment]
You're correct that there is no specific framework support for decrypting parts of a request / encrypting parts of a response.  However, it's possible to create your own 'wrapper' type that represents the encrypted request, with a decrypt function that decrypts the ciphertext, and then decodes that into the Codable type you expect.

If you want to send encrypted responses, you could make the wrapper work in both directions so that you can construct the wrapper using a Codable type.

Here's a little prototype that receives and responds with an encrypted payload: https://github.com/djones6/KituraEncrypt

**Comment by KaanYT** [comment]
Hi David,

Yes that should work but it is increasing redundancy especially in large project. I should be able to control it from one central point. For an enterprise ready product, I believe this should be implemented. Changeable RouterRequest body might become handy for other type of cases. And it gives flexibility.

**Comment by djones6** [comment]
I agree, it would be helpful for the request body to be mutated prior to being consumed by other handlers.  In the short term something like this _could_ be hacked into Kitura 2, but I'd see that as more of a stop-gap solution.

What I'd really like is to provide a mechanism where you stream the bytes of the request body, and can plug in middlewares that are essentially filters.  This is quite a significant architectural change from what Kitura 2 is today, but moving to a model that is much closer aligned with NIO's design is something we're really interested in doing for a Kitura 3.

In the meantime, feel free to reach out on our Slack for advice / code review if you'd like to pursue implementing a writable request body feature in Kitura 2.

**Comment by djones6** [comment]
Kitura already provides a `writtenDataFilter` which middlewares such as [Kitura-Compression](https://github.com/IBM-Swift/Kitura-Compression/) make use of.

There is currently no equivalent `readDataFilter`, but that's probably the right approach to apply transformations to the data as it is read in.

**Comment by ianpartridge** [comment]
I remember @shmuelk saying he thought the design of `writtenDataFilter` was a mistake and he'd change it if he had his time again... @shmuelk any thoughts on this?

**Comment by shmuelk** [comment]
Unfortunately, I haven't looked at this (or Swift) in quite a while.

I do remember that my design wasn't the greatest. I think I tried to emulate the Node.js "API" a little too closely.
